rotkehlchen/exchanges/bitcoinde.py

```diff
@@ -0,0 +1,249 @@
+                        raise RemoteError('Provided API Key is in invalid Format')
+
+                    if error.get('code') == 3:
+                        raise RemoteError('Provided API Key is invalid')
+
+                raise RemoteError(json_ret['errors'])
+
+            raise RemoteError(
+                'Bitcoin.de api request for {} failed with HTTP status code {}'.format(
+                    response.url,
+                    response.status_code,
+                ),
+            )
+
+        if not isinstance(json_ret, dict):
+            raise RemoteError('Bitcoin.de returned invalid non-dict response')
+
+        return json_ret
+
+    def validate_api_key(self) -> Tuple[bool, str]:
+        "Validates that the Bitcoin.de API key is good for usage in Rotki"
+
+        try:
+            self._api_query('get', 'account')
+            return True, ""
+
+        except RemoteError as e:
+            return False, str(e)
+
+    def query_balances(self, **kwargs: Any) -> Tuple[Optional[Dict[Asset, Dict[str, Any]]], str]:
+        balances = {}
+        resp_info = self._api_query('get', 'account')
+
+        for currency, balance in resp_info['data']['balances'].items():
+            asset = bitcoinde_asset(currency)
+            try:
+                usd_price = Inquirer().find_usd_price(asset=asset)
+            except RemoteError as e:
+                self.msg_aggregator.add_error(
+                    f'Error processing Bitcoin.de balance entry due to inability to '
+                    f'query USD price: {str(e)}. Skipping balance entry',
+                )
+                continue
+
+            balances[asset] = {
+                'amount': balance['total_amount'],
+                'usd_value': balance['total_amount'] * usd_price,
+            }
+
+        return balances, ""
+
+    def query_online_trade_history(
+            self,
+            start_ts: Timestamp,
+            end_ts: Timestamp,
+    ) -> List:
+
+        page = 1
+        resp_trades = []
+
+        while True:
+            resp = self._api_query('get', 'trades', {'state': 1, 'page': page})
+            resp_trades.extend(resp['trades'])
+
+            if 'page' not in resp:
+                break
+
+            if resp['page']['current'] >= resp['page']['last']:
+                break
+
+            page = resp['page']['current'] + 1
+
+        log.debug('Bitcoin.de trade history query', results_num=len(resp_trades))
+
+        trades = []
+        for tx in resp_trades:
+            try:
+                timestamp = iso8601ts_to_timestamp(tx['successfully_finished_at'])
+            except KeyError:
+                # For very old trades (2013) bitcoin.de does not return 'successfully_finished_at'
+                timestamp = iso8601ts_to_timestamp(tx['trade_marked_as_paid_at'])
+
+            if tx['state'] != 1:
+                continue
+            if timestamp and timestamp < start_ts:
```

Comment:
Why `if timestamp` here?

Is it just to silence mypy? You can probably do it by putting a `timestamp: Timestamp` above the for. If not just use` type: ignore comment`